// nuxt3/dist/pages/runtime/page.mjs
import { defineComponent, h, Suspense, Transition } from "vue";
import { RouterView } from "vue-router";
import { generateRouteKey, wrapIf, wrapInKeepAlive } from "./utils.mjs";
import { useNuxtApp } from "#app";
export default defineComponent({
  name: "NuxtPage",
  props: {
    pageKey: {
      type: [Function, String],
      default: null
    }
  },
  setup(props) {
    const nuxtApp = useNuxtApp();
    return () => {
      return h(RouterView, {}, {
        default: (routeProps) => routeProps.Component && wrapIf(Transition, routeProps.route.meta.pageTransition ?? defaultPageTransition, wrapInKeepAlive(routeProps.route.meta.keepalive, h(Suspense, {
          onPending: () => nuxtApp.callHook("page:start", routeProps.Component),
          onResolve: () => nuxtApp.callHook("page:finish", routeProps.Component)
        }, { default: () => h(routeProps.Component, { key: generateRouteKey(props.pageKey, routeProps) }) }))).default()
      });
    };
  }
});
const defaultPageTransition = { name: "page", mode: "out-in" };
